### ISSUE TITLE
usage of privacy and custom Go types result in circular import

### ISSUE THREAD

**Issue by DeedleFake** [ent/ent#923]
Given the following schema, there will be a circular import where `ent/schema` imports `ent/privacy` which imports `ent` which imports `ent/post` which imports `ent/schema`.

```go
type Post struct {
  ent.Schema
}

func (Post) Fields() []ent.Field {
  return []ent.Field{
    field.String("CustomType").
      GoType(new(SomeCustomType)),
  }
}

func (Post) Edges() []ent.Edge {
  return []ent.Edge{
    edge.From("User", User.Type).Ref("Posts").
      Unique(),
  }
}

func (Post) Policy() ent.Policy {
  return privacy.Policy{
    QueryPolicy: privacy.QueryPolicy{
      privacy.AlwaysAllowRule(),
    },
    MutationPolicy: privacy.MutationPolicy{
      privacy.AlwaysDenyRule(),
    },
  }
}

type User struct {
  ent.Schema
}

func (User) Edges() []ent.Edge {
  return []ent.Edge{
    edge.To("Posts", Post.Type),
  }
}
```

I was able to get around the issue by refactoring the custom types into a different package, but I thought it was worth noting because it was a bit awkward to track down what was causing it.

**Comment by masseelch** [comment]
I think you need to import the generated `ent/runtime` package like described [here](https://entgo.io/docs/privacy/#configuration).

```go
import _ "<project>/ent/runtime"
```

**Comment by DeedleFake** [comment]
I tried that. It doesn't help. The import cycle happens during the code generation, before the `main` package even gets involved.

**Comment by art-Hasan** [comment]
@DeedleFake are you able to sovle this problem?

**Comment by DeedleFake** [comment]
As I said at the bottom of the original post, hidden under the code block,

> I was able to get around the issue by refactoring the custom types into a different package, but I thought it was worth noting because it was a bit awkward to track down what was causing it.

In other words, I had to move all of the custom field types out of the `ent/schema` package and into their own package. That fixed it.

**Comment by a8m** [comment]
Hey @DeedleFake,
Someone mentioned this issue on our [Slack channel](https://entgo.io/docs/slack/), and I found out that I didn't reply to it or missed it somehow (Sorry!). 

When defining privacy rules on `ent/schema`, they usually import helpers from the generated `ent/privacy` package, which imports the `ent` package. Therefore, if you define a custom type on the `ent/schema` package and use it as a `GoType` in one of the fields, it will be used by the different generated builders/types, and this will cause a circular import (not only on the codegen phase). 

```
ent/schema -> ent/priacy -> ent -> ent/schema
```

The approach, as you mentioned, is to move the custom types into a separate package under the `ent/schema` package. For example `ent/schema/pulid` or `ent/schema/schematype`.

I can't think of a better approach for addressing this issue. We tried a few hacks to solve it 1 year ago, but all of them were too magic or didn't work well when the schema was complex. Therefore, I'm going to close this issue, but if someone has a creative suggestion on how to tackle this problem I'll be more than happy to reopen and address it. Thanks 🙏
